### PR TITLE
Model: Add Parent class

### LINKED ISSUE #67: Model: Add Parent Class
Abstract Parent details as its own class

### PR DESCRIPTION
**Pull request by IsaacPangTH**
Fixes #67 
Add Parent class to model. Additions not implemented.

### CODE REVIEW

File: src/main/java/seedu/address/model/person/Parent.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * Represents a parent in the address book
+ */
+public class Parent extends Person {
+    private Student child = null;
+
+    /**
+     * Constructs a {@code Parent} with the given details.
+     */
+    public Parent(Name name, Phone phone, Email email, Address address, Student child, Set<Tag> tags) {
+        super(name, phone, email, address, tags);
```
Comment (ckclion) [inline src/main/java/seedu/address/model/person/Parent.java:19]: Perhaps more tests could be written for this to hit minimum code coverage as determined by codecov!

File: src/main/java/seedu/address/model/person/Parent.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    /**
+     * Constructs a {@code Parent} with the given {@code Person} as a base.
+     */
+    public Parent(Person person, Student child) {
+        super(person.getName(), person.getPhone(), person.getEmail(), person.getAddress(), person.getTags());
+        this.child = child;
+    }
+
+    public void setChild(Student child) {
+        this.child = child;
```
Comment (ckclion) [inline src/main/java/seedu/address/model/person/Parent.java:32]: Same issue as above.

File: src/test/java/seedu/address/model/person/ParentTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
+package seedu.address.model.person;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static seedu.address.logic.commands.CommandTestUtil.VALID_ADDRESS_BOB;
+import static seedu.address.logic.commands.CommandTestUtil.VALID_EMAIL_BOB;
+import static seedu.address.logic.commands.CommandTestUtil.VALID_NAME_BOB;
+import static seedu.address.logic.commands.CommandTestUtil.VALID_PHONE_BOB;
+import static seedu.address.logic.commands.CommandTestUtil.VALID_TAG_HUSBAND;
```
Comment (ckclion) [inline src/test/java/seedu/address/model/person/ParentTest.java:10]: I like that the imports are explicitly listed

File: src/test/java/seedu/address/testutil/TypicalParents.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import seedu.address.model.AddressBook;
+import seedu.address.model.person.Parent;
+
+/**
+ * A utility class containing a list of {@code Person} objects to be used in tests.
```
Comment (ckclion) [inline src/test/java/seedu/address/testutil/TypicalParents.java:22]: Perhaps putting {@code Parent} would be more appropriate here?

File: src/test/java/seedu/address/testutil/TypicalParents.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            .withEmail(VALID_EMAIL_BOB).withAddress(VALID_ADDRESS_BOB).withTags(VALID_TAG_HUSBAND, VALID_TAG_FRIEND)
+            .build();
+
+    public static final String KEYWORD_MATCHING_MEIER = "Meier"; // A keyword that matches MEIER
+
+    private TypicalParents() {
+    } // prevents instantiation
+
+    /**
+     * Returns an {@code AddressBook} with all the typical persons.
```
Comment (ckclion) [inline src/test/java/seedu/address/testutil/TypicalParents.java:85]: Same issue here.

**Review by ckclion — COMMENTED** [review]
Just some nitpicks with javadoc comments and more tests needed to satisfy codecov minimum coverage, otherwise LGTM!

File: src/test/java/seedu/address/testutil/ParentBuilder.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+     * Initializes the ParentBuilder with the data of {@code parentToCopy}.
+     */
+    public ParentBuilder(Parent parentToCopy) {
+        super(parentToCopy);
+        child = parentToCopy.getChild();
+    }
+
+    /**
+     * Sets the {@code child} of the {@code Parent} that we are building.
+     */
```
Comment (elliot-tang) [inline src/test/java/seedu/address/testutil/ParentBuilder.java:35]: I like how the JavaDoc comments are clear and concise, and have explicit links to the relevant classes!

File: src/test/java/seedu/address/testutil/ParentBuilder.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    public ParentBuilder withChild(Student child) {
+        this.child = child;
+        return this;
+    }
+
+    @Override
```
Comment (elliot-tang) [inline src/test/java/seedu/address/testutil/ParentBuilder.java:41]: Perhaps adding similar JavaDoc comments to the remaining withXXX() methods would help with readability and navigability to relevant classes!

File: src/test/java/seedu/address/model/person/ParentTest.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    public void equals() {
+        // same values -> returns true
+        Person aliceCopy = new ParentBuilder(ALICE).build();
+        assertTrue(ALICE.equals(aliceCopy));
+
+        // same object -> returns true
+        assertTrue(ALICE.equals(ALICE));
+
+        // null -> returns false
+        assertFalse(ALICE.equals(null));
```
Comment (elliot-tang) [inline src/test/java/seedu/address/model/person/ParentTest.java:65]: Great that even the individual code segments within the equals method are clearly explained with code comments!

File: src/main/java/seedu/address/model/person/Parent.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        Parent otherParent = (Parent) other;
+        return super.equals(otherParent) && getChild().equals(otherParent.getChild());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.getName(), this.getPhone(), this.getEmail(), this.getAddress(),
+                this.getChild(), this.getTags());
```
Comment (elliot-tang) [inline src/main/java/seedu/address/model/person/Parent.java:57]: Perhaps you could consider adding a test for the hashCode method? You could create a test case in which you obtain the individual fields manually, compute the hash and use assertEquals() to compare the resultant value with hashCode()

File: src/test/java/seedu/address/model/person/ParentTest.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+import seedu.address.testutil.ParentBuilder;
+
+public class ParentTest {
+
+    @Test
+    public void asObservableList_modifyList_throwsUnsupportedOperationException() {
+        Person person = new ParentBuilder().build();
+        assertThrows(UnsupportedOperationException.class, () -> person.getTags().remove(0));
+    }
```
Comment (elliot-tang) [inline src/test/java/seedu/address/model/person/ParentTest.java:26]: I like that the test method name is clear and follows the established test method naming conventions!

**Review by elliot-tang — APPROVED** [review]
LGTM! Perhaps you can consider adding several more test cases to improve the code coverage.

File: src/test/java/seedu/address/model/person/ParentTest.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    public void toMessageString() {
+        StringBuilder tags = new StringBuilder();
+        ALICE.getTags().forEach(tags::append);
+        String expected = ALICE.getName() + "; Phone: " + ALICE.getPhone() + "; Email: " + ALICE.getEmail()
+                + "; Address: " + ALICE.getAddress() + "; Child: " + ALICE.getChild() + "; Tags: " + tags;
+        assertEquals(expected, ALICE.toMessageString());
+    }
+
+    @Test
+    public void toStringMethod() {
```
Comment (oneBoz) [inline src/test/java/seedu/address/model/person/ParentTest.java:108]: I like that the naming is consistent throughout the different tests

File: src/test/java/seedu/address/model/person/ParentTest.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        // same name, all other attributes different -> returns true
+        Person editedAlice = new ParentBuilder(ALICE).withPhone(VALID_PHONE_BOB).withEmail(VALID_EMAIL_BOB)
+                .withAddress(VALID_ADDRESS_BOB).withTags(VALID_TAG_HUSBAND).withChild(CARL).build();
+        assertTrue(ALICE.isSamePerson(editedAlice));
+
+        // different name, all other attributes same -> returns false
+        editedAlice = new ParentBuilder(ALICE).withName(VALID_NAME_BOB).build();
+        assertFalse(ALICE.isSamePerson(editedAlice));
+
+        // name differs in case, all other attributes same -> returns false
```
Comment (oneBoz) [inline src/test/java/seedu/address/model/person/ParentTest.java:45]: I like how the testing is comprehensive and covers various edge cases

File: src/test/java/seedu/address/model/person/ParentTest.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+        // different child -> returns false
+        editedAlice = new ParentBuilder(ALICE).withChild(CARL).build();
+        assertFalse(ALICE.equals(editedAlice));
+    }
+
+    @Test
+    public void toMessageString() {
+        StringBuilder tags = new StringBuilder();
+        ALICE.getTags().forEach(tags::append);
```
Comment (oneBoz) [inline src/test/java/seedu/address/model/person/ParentTest.java:101]: I like how you clean this line looks

**Review by oneBoz — COMMENTED** [review]
LGTM! Perhaps more test cases need to be added. Really well designed code!

File: src/test/java/seedu/address/testutil/TypicalParents.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
+package seedu.address.testutil;
+
+import static seedu.address.logic.commands.CommandTestUtil.VALID_ADDRESS_AMY;
+import static seedu.address.logic.commands.CommandTestUtil.VALID_ADDRESS_BOB;
+import static seedu.address.logic.commands.CommandTestUtil.VALID_EMAIL_AMY;
+import static seedu.address.logic.commands.CommandTestUtil.VALID_EMAIL_BOB;
```
Comment (lutfir-cpu) [inline src/test/java/seedu/address/testutil/TypicalParents.java:6]: good that import statements follow code standards, and without the use of wildcards.

**Review by lutfir-cpu — APPROVED** [review]
LGTM, but more test cases could be included for more comprehensive testing.
